torch/csrc/TensorOptions.h

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <torch/csrc/autograd/generated/VariableType.h>
+#include <torch/csrc/autograd/variable.h>
+
+#include <ATen/Error.h>
+#include <ATen/Tensor.h>
+#include <ATen/TensorOptions.h>
+#include <ATen/Type.h>
+
+namespace torch {
+struct TensorOptions : public at::TensorOptions {
```

Comment:
we're going in the direction of folding autograd into core Tensor class. If that's the case - shall requires_grad be present just in the parent class so that we can avoid 2 different types?

Reply:
Though then apply() won't work..

Reply:
Yes `torch::TensorOptions` exists mainly to solve the problem of setting `requires_grad` for new variables, which I've achieved via polymorphism here. I figured it's better to only allow setting `requires_grad` when using this subclass, then. But it could just as well be in ATen `TensorOptions`

Reply:
My concern with this subclassing is that we are making something that frequently will just be constant-propagated away into a bunch of virtual function calls. Can we make requires_grad an optional<bool>?

true - Variable requires grad
false - Variable no grad
none - tensor

Like at::CPU vs torch::CPU we can make the default based on which function you used to create the TensorOption.